File: common/src/main/java/net/shiroha233/roadweaver/planning/impl/MSTPlanner.java
package net.shiroha233.roadweaver.planning.impl;

import net.minecraft.core.BlockPos;
import net.shiroha233.roadweaver.core.model.StructureConnection;
import net.shiroha233.roadweaver.planning.NetworkPlanner;
import net.shiroha233.roadweaver.planning.PlanningUtils;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;

/**
 * 基于最小生成树（Kruskal）的路网规划器
 */
public final class MSTPlanner implements NetworkPlanner {

    @Override
    public List<StructureConnection> plan(List<BlockPos> points, int maxEdgeLenBlocks) {
        return planMST(points, maxEdgeLenBlocks);
    }

    public static List<StructureConnection> planMST(List<BlockPos> points, int maxEdgeLenBlocks) {
        if (points == null || points.size() < 2) return List.of();

        ArrayList<BlockPos> unique = new ArrayList<>();
        HashSet<Long> seen = new HashSet<>();
        for (BlockPos p : points) {
            BlockPos q = new BlockPos(p.getX(), 0, p.getZ());
            long key = PlanningUtils.pos2dKey(q);
            if (seen.add(key)) unique.add(q);
        }
        int n = unique.size();
        if (n < 2) return List.of();

        long maxD2 = maxEdgeLenBlocks > 0 ? (long) maxEdgeLenBlocks * (long) maxEdgeLenBlocks : Long.MAX_VALUE;

        record Edge(int a, int b, long d2) {}

        ArrayList<Edge> edges = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            BlockPos pi = unique.get(i);
            long ix = pi.getX();
            long iz = pi.getZ();
            for (int j = i + 1; j < n; j++) {
                BlockPos pj = unique.get(j);
                long dx = ix - pj.getX();
                long dz = iz - pj.getZ();
                long d2 = dx * dx + dz * dz;
                if (d2 > maxD2) continue;
                edges.add(new Edge(i, j, d2));
            }
        }
        if (edges.isEmpty()) return List.of();

        edges.sort((e1, e2) -> Long.compare(e1.d2(), e2.d2()));

        int[] parent = new int[n];
        for (int i = 0; i < n; i++) parent[i] = i;

        ArrayList<StructureConnection> result = new ArrayList<>();
        HashSet<Long> edgeKeys = new HashSet<>();

        for (Edge e : edges) {
            int ra = find(parent, e.a());
            int rb = find(parent, e.b());
            if (ra == rb) continue;
            parent[rb] = ra;

            int ia = Math.min(e.a(), e.b());
            int ib = Math.max(e.a(), e.b());
            long key = (((long) ia) << 32) ^ (long) ib;
            if (!edgeKeys.add(key)) continue;

            BlockPos pa = unique.get(e.a());
            BlockPos pb = unique.get(e.b());
            result.add(new StructureConnection(pa, pb));
        }

        return result;
    }

    private static int find(int[] parent, int x) {
        while (parent[x] != x) {
            parent[x] = parent[parent[x]];
            x = parent[x];
        }
        return x;
    }
}

File: common/src/main/kotlin/net/shiroha233/roadweaver/planning/MSTPlanner.kt
package net.shiroha233.roadweaver.planning

import net.minecraft.core.BlockPos
import net.shiroha233.roadweaver.helpers.Records
import java.util.ArrayList
import java.util.HashSet

object MSTPlanner {
    @JvmStatic
    fun planMST(points: List<BlockPos>?, maxEdgeLenBlocks: Int): List<Records.StructureConnection> {
        if (points === null || points.size < 2) return listOf()

        val unique = ArrayList<BlockPos>()
        val seen = HashSet<Long>()
        for (p in points) {
            val q = BlockPos(p.x, 0, p.z)
            val key = PlanningUtils.pos2dKey(q)
            if (seen.add(key)) unique.add(q)
        }
        val n = unique.size
        if (n < 2) return listOf()

        val maxD2 = if (maxEdgeLenBlocks > 0) {
            maxEdgeLenBlocks.toLong() * maxEdgeLenBlocks.toLong()
        } else {
            Long.MAX_VALUE
        }

        data class Edge(val a: Int, val b: Int, val d2: Long)

        val edges = ArrayList<Edge>()
        for (i in 0 until n) {
            val pi = unique[i]
            val ix = pi.x.toLong()
            val iz = pi.z.toLong()
            for (j in i + 1 until n) {
                val pj = unique[j]
                val dx = ix - pj.x.toLong()
                val dz = iz - pj.z.toLong()
                val d2 = dx * dx + dz * dz
                if (d2 > maxD2) continue
                edges.add(Edge(i, j, d2))
            }
        }
        if (edges.isEmpty()) return listOf()

        edges.sortWith(compareBy<Edge> { it.d2 })

        val parent = IntArray(n) { it }

        val result = ArrayList<Records.StructureConnection>()
        val edgeKeys = HashSet<Long>()

        for (e in edges) {
            val ra = find(parent, e.a)
            val rb = find(parent, e.b)
            if (ra == rb) continue
            parent[rb] = ra

            val ia = minOf(e.a, e.b)
            val ib = maxOf(e.a, e.b)
            val key = (ia.toLong() shl 32) xor ib.toLong()
            if (!edgeKeys.add(key)) continue

            result.add(Records.StructureConnection(unique[e.a], unique[e.b]))
        }

        return result
    }

    private fun find(parent: IntArray, x0: Int): Int {
        var x = x0
        while (parent[x] != x) {
            parent[x] = parent[parent[x]]
            x = parent[x]
        }
        return x
    }
}
